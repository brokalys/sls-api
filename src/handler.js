import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (process.env.STAGE !== 'prod' && process.env.NODE_ENV !== 'test') ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (process.env.NODE_ENV !== 'test') {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
