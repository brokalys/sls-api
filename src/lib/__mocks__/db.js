export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
