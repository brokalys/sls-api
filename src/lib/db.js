import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
