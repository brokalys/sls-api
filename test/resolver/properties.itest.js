import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';
import makeProperty from '../db/fixtures/property';

const { query } = createTestClient(server);

describe('Query: properties', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  test('successfully retrieves property summary.count data', async () => {
    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              count
            }
          }
        }
      `,
      variables: {
        filter: {
          source: { eq: 'ss.lv' },
          type: { eq: 'rent' },
        },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves all the property fields', async () => {
    authenticateAs('slsCrawler', server);

    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            results {
              id
              url
              category
              type
              rent_type
              content
              images
              price
              calc_price_per_sqm
              rooms
              area
              floor
              lat
              lng
              published_at
              foreign_id
            }
            summary {
              count
              price {
                count
                min
                max
                mean
                median
                mode
                standardDev
              }
            }
          }
        }
      `,
      variables: {
        filter: {
          source: { eq: 'ss.lv' },
          type: { eq: 'rent' },
        },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('returns 0 count when no properties are found', async () => {
    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              count
            }
          }
        }
      `,
      variables: {
        filter: { price: { gt: 9999999 } },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves property summary.price data', async () => {
    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              price {
                count
                min
                max
                mean
                median
                mode
                standardDev
              }
            }
          }
        }
      `,
      variables: {
        filter: {
          source: { eq: 'ss.lv' },
          type: { eq: 'rent' },
        },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves truncated property summary.price data', async () => {
    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              price(discard: 0.5) {
                count
                min
                max
                mean
                median
                mode
                standardDev
              }
            }
          }
        }
      `,
      variables: {
        filter: { source: { eq: 'ss.lv' } },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test('successfully retrieves results', async () => {
    authenticateAs('slsDataExtraction', server);

    const response = await query({
      query: `
       query ($filter: PropertyFilter) {
          properties(filter: $filter) {
            results {
              id
            }
          }
        }
      `,
      variables: {
        filter: { category: { eq: 'apartment' } },
      },
    });

    expect(response).toMatchSnapshot();
  });

  test.each([
    { filter: { published_at: { gte: 'value' } } }, // simply wrong value
    { filter: { published_at: '2019-01-01' } }, // wrong filter expression
  ])(
    'throws a validation error when trying to retrieve summary.count',
    async (input) => {
      const response = await query({
        query: `
        query GetCount($filter: PropertyFilter) {
          properties(filter: $filter) {
            summary {
              count
            }
          }
        }
      `,
        variables: input,
      });

      expect(response.errors).toHaveLength(1);
    },
  );

  test('throws an authentication error if trying to retrieve results without authorizing', async () => {
    const response = await query({
      query: `
        {
          properties {
            results {
              id
            }
          }
        }
      `,
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });

  test('should not be able to query unlimited results with public permissions', async () => {
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        {
          properties(limit: null) {
            results {
              id
            }
          }
        }
      `,
    });

    expect(response.errors).toEqual([
      expect.objectContaining({
        extensions: {
          code: 'UNAUTHENTICATED',
        },
      }),
    ]);
  });

  test('should limit the results to 30 if no limit provided', async () => {
    await db('properties').insert(new Array(100).fill(makeProperty()));
    authenticateAs('mapApp', server);

    const response = await query({
      query: `
        {
          properties {
            results {
              id
            }
          }
        }
      `,
    });

    expect(response.data.properties.results.length).toBe(20);
  });

  test.each(['url', 'content', 'images', 'lat', 'lng', 'foreign_id'])(
    'throws an authentication error if trying to retrieve specific field %j without permissions',
    async (field) => {
      authenticateAs('mapApp', server);

      const response = await query({
        query: `
        {
          properties {
            results {
              ${field}
            }
          }
        }
      `,
      });

      expect(response.errors).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            extensions: {
              code: 'UNAUTHENTICATED',
            },
          }),
        ]),
      );
    },
  );
});
