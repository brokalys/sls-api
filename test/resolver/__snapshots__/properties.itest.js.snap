// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query: properties returns 0 count when no properties are found 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "count": 0,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves all the property fields 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "area": 90,
          "calc_price_per_sqm": 1.11,
          "category": "apartment",
          "content": "",
          "floor": 2,
          "foreign_id": null,
          "id": 1,
          "images": Array [],
          "lat": 56.1,
          "lng": 24.9,
          "price": 100,
          "price_per_sqm": 1.11,
          "published_at": "2021-04-12T19:47:30.000Z",
          "rent_type": "monthly",
          "rooms": null,
          "type": "rent",
          "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
        },
        Object {
          "area": null,
          "calc_price_per_sqm": null,
          "category": "apartment",
          "content": "",
          "floor": null,
          "foreign_id": null,
          "id": 2,
          "images": Array [],
          "lat": 56.1,
          "lng": 24.9,
          "price": 200,
          "price_per_sqm": null,
          "published_at": "2021-04-12T19:47:30.000Z",
          "rent_type": "monthly",
          "rooms": null,
          "type": "rent",
          "url": "https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html",
        },
        Object {
          "area": null,
          "calc_price_per_sqm": null,
          "category": "house",
          "content": "",
          "floor": null,
          "foreign_id": null,
          "id": 3,
          "images": Array [],
          "lat": null,
          "lng": null,
          "price": 300,
          "price_per_sqm": null,
          "published_at": "2021-04-12T19:47:30.000Z",
          "rent_type": "monthly",
          "rooms": null,
          "type": "rent",
          "url": "https://www.ss.com/msg/lv/real-estate/homes-summer-residences/tukums-and-reg/engures-pag/hhjbf.html",
        },
      ],
      "summary": Object {
        "count": 3,
        "price": Object {
          "count": 3,
          "max": 300,
          "mean": 200,
          "median": 200,
          "min": 100,
          "mode": 100,
          "standardDev": 81,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves property summary.count data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "count": 3,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves property summary.price data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "price": Object {
          "count": 3,
          "max": 300,
          "mean": 200,
          "median": 200,
          "min": 100,
          "mode": 100,
          "standardDev": 81,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves results 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: properties successfully retrieves truncated property summary.price data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "price": Object {
          "count": 2,
          "max": 300,
          "mean": 250,
          "median": 250,
          "min": 200,
          "mode": 200,
          "standardDev": 50,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
